### PR TITLE
WIP: Add scripts for bulk annotation 

### PR DESCRIPTION
**Pull request by CFGrote**
This PR adds scripts originally maintained by @evenhuis to support bulk annotation via csv files and dataset descriptions. The original code was ported to python3, bugs fixed and rudimentary tests added. 

Marked as Draft here and seeking feedback and contributions from @will-moore , @ahamacher and others as discussed in [this images.sc thread](https://forum.image.sc/t/uploading-key-value-pairs-from-csv-files-into-omero-web-for-plates/60202/23).

----

Update: Please see #193

### CODE REVIEW

File: omero/annotation_scripts/KeyVal_from_csv.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                        for val in vals:
+                            if len(val)>0 : 
+                                if key not in updated_kv: updated_kv[key] = set()
+                                print("adding",key,val)
+                                updated_kv[key].add(val)
+
+                if( existing_kv != updated_kv ):
+                    nimg_updated = nimg_updated + 1
+                    print("The key-values pairs are different")
+                    remove_MapAnnotations( conn, 'Image', img.getId()  )
```
Comment (will-moore) [inline omero/annotation_scripts/KeyVal_from_csv.py]: This is named incorrectly - should be `remove_map_annotations`.

File: omero/annotation_scripts/KeyVal_from_csv.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+from omero.util.populate_roi import DownloadingOriginalFileProvider
+from omero.util.populate_metadata import ParsingContext
+
+from collections import OrderedDict
+
+
+
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+def get_existing_map_annotions( obj ):
```
Comment (will-moore) [inline omero/annotation_scripts/KeyVal_from_csv.py]: Typo: should be `get_existing_map_annotations`

File: omero/annotation_scripts/KeyVal_from_csv.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            "IDs", optional=False, grouping="2",
+            description="Plate or Screen ID.").ofType(rlong(0)),
+
+        scripts.String(
+            "File_Annotation", grouping="3",
+            description="File ID containing metadata to populate."),
+
+        authors=["Christian Evenhuis"],
+        institutions=["MIF UTS"],
+        contact="christian.evenhuis@gmail.com"
```
Comment (will-moore) [inline omero/annotation_scripts/KeyVal_from_csv.py:286]: Update the authors and maybe institutions in all scripts to add yourself and OME team?
But certainly update the contact in all scripts, probably to `https://forum.image.sc/tag/omero`.  cc @joshmoore ?

**Review by will-moore — CHANGES_REQUESTED** [review]
Added some initial comments...

File: omero/annotation_scripts/KeyVal_from_csv.py (hunk 3 above)
Comment (joshmoore) [reply omero/annotation_scripts/KeyVal_from_csv.py:286]: That tag is certainly fine thought we could also create an `omero-scripts` or similar if there's interest.

**Comment by joshmoore** [comment]
Since it's happened to me on other repos, I took the time to at least _explore_ saving the history in https://github.com/ome/omero-scripts/pull/192

**Comment by will-moore** [comment]
@CFGrote Just looking to test the various scripts added here, using your blog at https://mpievolbio-scicomp.pages.gwdg.de/blog/post/2020-09-03_omerobulkannotation/ as a guide.
I notice that doesn't mention `KeyVal from Description`, `KeyVal from Filename` or `Remove KeyVal` and I can't guess from the script dialog description how to use them. Do you use these scripts?
I'm wondering if we really want to include them here; Is the cost of maintenance, testing etc justified?

**Comment by CFGrote** [comment]
yes, sorry, those scripts are not mentioned. 

https://github.com/mpievolbio-scicomp/ometa/blob/master/readme.orig.md has more information and links to 
more detailed documentation at https://code.research.uts.edu.au/MIF/OMERO-instructions.

KeyVal_from_Description: Parses the Description field of a Dataset and converts "key: val" strings to key-value annotations of the same Dataset (if I remember correctly). We used this for a little while before getting KeyVal_from_csv to work for us. I'd say it can be removed.

KeyVal_from_Filename: Never used it. It should tokenize image filenames into key-value pairs based on a pattern defined again in the Dataset description. Can be removed because that functionality is provided in Insight (correct?).

RemoveKeyVal: Strips off all key-val pairs in the special namespace from selected items. Issue is that it does not really remove the annotations (as in "delete from the database") but somehow hides from the view. They resurface under certain conditions like applying KeyVal_from_csv again.  This one should stay and be debugged IMHO.

Hope that helps.

Thanks and best regards, Carsten

**Comment by joshmoore** [comment]
Closing in favor of #193

**Comment by will-moore** [comment]
@CFGrote I think KeyVal_from_Filename functionality is provided by the auto-tagging omero.web plugin to create Tags rather than KV pairs. Certainly if you've never used it then let's keep thing simpler by not including it.

I've created an issue for RemoveKeyVal https://github.com/ome/omero-scripts/issues/194

Moving this discussion to #193...

**Comment by CFGrote** [comment]
yes, you're right. sorry for the confusion.

**Pull request closed without merging**